tests/usecases/grant_permission_to_group_test.py

```diff
@@ -0,0 +1,198 @@
+    usecase.grant_permission_to_group("some-permission", "argument", "some-group")
+    assert mock_ui.mock_calls == [
+        call.grant_permission_to_group_failed_permission_already_exists("some-group")
+    ]
+
+
+def test_invalid_argument(setup):
+    # type: (SetupTest) -> None
+    with setup.transaction():
+        setup.add_user_to_group("gary@a.co", "admins")
+        setup.grant_permission_to_group(PERMISSION_ADMIN, "", "admins")
+        setup.create_service_account("service@svc.localhost", "some-group")
+        setup.create_permission("some-permission")
+
+    mock_ui = MagicMock()
+    usecase = setup.usecase_factory.create_grant_permission_to_group_usecase("gary@a.co", mock_ui)
+    usecase.grant_permission_to_group("some-permission", "@@@@", "some-group")
+    assert mock_ui.mock_calls == [
+        call.grant_permission_to_group_failed_invalid_argument(
+            "some-permission",
+            "@@@@",
+            "some-group",
+            "Permission argument is not valid (does not match {})".format(ARGUMENT_VALIDATION),
+        )
+    ]
+
+
+def _assert_sane_permission_denied_message(mock_ui):
+    # type: (MagicMock) -> None
+    # Check the mock's first call ([0]) non-keyword args ([1]) last arg ([-1]),
+    # which in permission denied cases is the error message.
+    assert mock_ui.mock_calls[0][1][-1].startswith("Permission denied")
+
+
+def test_permission_denied(setup):
+    # type: (SetupTest) -> None
+    with setup.transaction():
+        setup.create_user("gary@a.co")
+        setup.create_service_account("service@svc.localhost", "some-group")
+        setup.create_permission("some-permission")
+        setup.grant_permission_to_group("other-permission", "*", "some-group")
+
+    # User with no special permissions.
+    mock_ui = MagicMock()
+    usecase = setup.usecase_factory.create_grant_permission_to_group_usecase("gary@a.co", mock_ui)
+    usecase.grant_permission_to_group("some-permission", "argument", "some-group")
+    assert mock_ui.mock_calls == [
+        call.grant_permission_to_group_failed_permission_denied(
+            "some-permission", "argument", "some-group", ANY
+        )
+    ]
+    _assert_sane_permission_denied_message(mock_ui)
+
+    # Service account with no special permissions.
+    mock_ui.reset_mock()
+    usecase = setup.usecase_factory.create_grant_permission_to_group_usecase(
+        "service@svc.localhost", mock_ui
+    )
+    usecase.grant_permission_to_group("some-permission", "argument", "some-group")
+    assert mock_ui.mock_calls == [
+        call.grant_permission_to_group_failed_permission_denied(
+            "some-permission", "argument", "some-group", ANY
+        )
+    ]
+    _assert_sane_permission_denied_message(mock_ui)
+
+    # Add the user to the group, but don't delegate the permission to the group.  The user should
+    # now be able to grant permissions in general, but not this permission in particular.
```

Comment:
I think some of these comments are from the grant to service account usecase and don't apply here? E.g. being a member of the group in this case does not confer the ability to grant any permissions whatsoever